hera_cal/abscal.py

```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+    else:  # skip this step for speed
+        return all_times, all_lsts
+
+
+def get_d2m_time_map(data_times, data_lsts, model_times, model_lsts):
+    '''Generate a dictionary that maps data times to model times via shared LSTs.
+
+    Arguments:
+        data_times: list of times in the data (in JD)
+        data_lsts: list of corresponding LSTs (in radians)
+        model_times: list of times in the mdoel (in JD)
+        model_lsts: list of corresponing LSTs (in radians)
+
+    Returns:
+        d2m_time_map: dictionary uniqely mapping times in the data to times in the model 
+            that are closest in LST. Each model time maps to at most one data time and 
+            each model time maps to at most one data time. Data times without corresponding
+            model times map to None.
+    '''
+    # first produce a map of indices using the LSTs
+    m2d_ind_map = {}  
+    for dind, dlst in enumerate(data_lsts):
+        nearest_mind = np.argmin(np.abs(model_lsts - dlst))
+        if nearest_mind in m2d_ind_map:
+            if np.abs(model_lsts[nearest_mind] < data_lsts[m2d_ind_map[nearest_mind]]):
+                m2d_ind_map[nearest_mind] = dind
+        else:
+            m2d_ind_map[nearest_mind] = dind
+
+    # now use those indicies to produce a map of times
+    d2m_time_map = {time: None for time in data_times}
+    for mind, dind in m2d_ind_map.items():
+        d2m_time_map[data_times[dind]] = model_times[mind]
+    return d2m_time_map
+
+
+def abscal_step(gains_to_update, AC, AC_func, AC_kwargs, gain_funcs, gain_args_list, gain_flags, 
+                gain_convention='divide', max_iter=1, phs_conv_crit=1e-6, verbose=True):
+    '''Generalized function for performing an abscal step (e.g. abs_amp_logcal or TT_phs_logcal).
+
+    Arguments:
+        gains_to_update: the gains produced by abscal up until this step. Updated in place.
+        AC: AbsCal object containing data, model, and other metadata. AC.data is recalibrated 
+            in place using the gains solved for during this step
+        AC_func: function (usually a class method of AC) to call to instantiate the new gains 
+            which are then accessible as class properties of AC
+        AC_kwargs: dictionary of kwargs to pass into AC_func
+        gain_funcs: list of functions to call to return gains after AC_func has been called
+        gain_args_list: list of tuples of arguments to pass to the corresponding gain_funcs
+        gain_flags: per-antenna flags to apply to AC.Data when performing recalibration
+        gain_convention: either 'divide' if raw data is calibrated by dividing it by the gains
+            otherwise, 'multiply'.
+        max_iter: maximum number of times to run phase solvers iteratively to avoid the effect
+            of phase wraps in, e.g. phase_slope_cal or TT_phs_logcal
+        phs_conv_crit: convergence criterion for updates to iterative phase calibration that compares
```

Comment:
function isn't specific to phs, so should be `conv_crit`

Reply:
Actually, the way I wrote it, it's very specific to phase. The non-phase steps don't suffer from phase wraps, which was the original motivation for making them iterative (unlike redcal, where the iterative steps address the fundamentally non-linear nature of redcal).